L1Trigger/L1TTrackMatch/src/L1TkMuCorrDynamicWindows.cc

```diff
@@ -0,0 +1,403 @@
+      throw cms::Exception("L1TkMuCorrDynamicWindows")
+          << "TF1 named " << nml << " or " << nmh << " not found in file " << fIn_theta->GetName() << ".\n";
+    wdws_theta_S1_.at(ib).SetName(wdn);
+    wdws_theta_S1_.at(ib).SetLower(fl);
+    wdws_theta_S1_.at(ib).SetUpper(fh);
+
+    wdn = std::string("wdw_phi_") + std::to_string(ib + 1);
+    nml = std::string("fit_low_") + std::to_string(ib + 1);
+    nmh = std::string("fit_high_") + std::to_string(ib + 1);
+    fl = (TF1*)fIn_phi_S1->Get(nml.c_str());
+    fh = (TF1*)fIn_phi_S1->Get(nmh.c_str());
+    if (fl == nullptr || fh == nullptr)
+      throw cms::Exception("L1TkMuCorrDynamicWindows")
+          << "TF1 named " << nml << " or " << nmh << " not found in file " << fIn_theta->GetName() << ".\n";
+    wdws_phi_S1_.at(ib).SetName(wdn);
+    wdws_phi_S1_.at(ib).SetLower(fl);
+    wdws_phi_S1_.at(ib).SetUpper(fh);
+  }
+}
+
+int L1TkMuCorrDynamicWindows::findBin(double val) {
+  // not the most efficient, nor the most elegant implementation for now
+  if (val < bounds_.at(0))
+    return 0;
+  if (val >= bounds_.back())
+    return (nbins_ - 1);  // i.e. bounds_size() -2
+
+  for (uint ib = 0; ib < bounds_.size() - 1; ++ib) {
+    if (val >= bounds_.at(ib) && val < bounds_.at(ib + 1))
+      return ib;
+  }
+
+  //"Something strange happened at val.
+  throw cms::Exception("L1TkMuCorrDynamicWindows") << "Can't find bin.\n";
+  return 0;
+}
+
+std::vector<int> L1TkMuCorrDynamicWindows::find_match(const EMTFTrackCollection& l1mus,
+                                                      const L1TTTrackCollectionType& l1trks) {
+  std::vector<int> out(l1trks.size());
+  for (auto l1trkit = l1trks.begin(); l1trkit != l1trks.end(); ++l1trkit) {
+    float trk_pt = l1trkit->momentum().perp();
+    float trk_p = l1trkit->momentum().mag();
+    float trk_aeta = std::abs(l1trkit->momentum().eta());
+    float trk_theta = to_mpio2_pio2(eta_to_theta(l1trkit->momentum().eta()));
+    float trk_phi = l1trkit->momentum().phi();
+    int trk_charge = (l1trkit->rInv() > 0 ? 1 : -1);
+
+    // porting some selections from the MuonTrackCorr finder
+    // https://github.com/cms-l1t-offline/cmssw/blob/l1t-phase2-932-v1.6/L1Trigger/L1TTrackMatch/plugins/L1TkMuonProducer.cc#L264
+    // in future, make preselections confiuguable
+    bool reject_trk = false;
+    if (trk_p < min_trk_p_)
+      reject_trk = true;
+    if (trk_aeta > max_trk_aeta_)
+      reject_trk = true;
+    if (track_qual_presel_) {
+      float l1tk_chi2 = l1trkit->chi2();
+      int l1tk_nstubs = l1trkit->getStubRefs().size();
+      if (l1tk_chi2 >= max_trk_chi2_)
+        reject_trk = true;
+      if (l1tk_nstubs < min_trk_nstubs_)
+        reject_trk = true;
+    }
+
+    int ibin = findBin(trk_aeta);
+
+    std::vector<std::tuple<float, float, int>> matched;  // dtheta, dphi, idx
+    // loop on muons to see which match
+    for (auto l1muit = l1mus.begin(); l1muit != l1mus.end(); ++l1muit) {
```

Comment:
Since there is no difference in performance, I prefer iterator type loop here because I use distance function inside the loop.

Reply:
Using a distance calculation inside a loop is a valid reason to keep the iterator-based loop.

"No performance difference" is not a valid reason, because in fact there is a performance difference: range-based loops avoid creating iterator temporaries, and are therefore faster. (It takes some tight nested loops to measure this effect, but it's there.)